go-controller/pkg/ovn/egressip.go

```diff
@@ -1996,9 +2014,118 @@ func createDefaultNoReroutePodPolicies(nbClient libovsdbclient.Client, v4Cluster
+	qoses := []*nbdb.QoS{}
+	if as, err = addressSetFactory.GetAddressSet(dbIDs); err != nil {
+		return nil, nil, fmt.Errorf("cannot ensure that addressSet %s exists %v", EgressIPServedPodsAddrSetName, err)
+	}
+	ipv4EgressIPServedPodsAS, ipv6EgressIPServedPodsAS := as.GetASHashNames()
+	qosRule := nbdb.QoS{
+		Priority:  types.EgressIPRerouteQoSRulePriority,
+		Action:    map[string]int{"mark": types.EgressIPReplyTrafficConnectionMark},
+		Direction: nbdb.QoSDirectionFromLport,
+	}
+	if config.IPv4Mode {
+		qosV4Rule := qosRule
+		qosV4Rule.Match = fmt.Sprintf(`ip4.src == $%s && ct.trk && ct.rpl`, ipv4EgressIPServedPodsAS)
+		qosV4Rule.ExternalIDs = getEgressIPQoSRuleDbIDs(IPFamilyValueV4).GetExternalIDs()
+		ops, err = libovsdbops.CreateOrUpdateQoSesOps(nbClient, nil, &qosV4Rule)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot create v4 QoS rule ops for egressIP feature on controller %s, %v", controllerName, err)
+		}
+		qoses = append(qoses, &qosV4Rule)
+	}
+	if config.IPv6Mode {
+		qosV6Rule := qosRule
+		qosV6Rule.Match = fmt.Sprintf(`ip6.src == $%s && ct.trk && ct.rpl`, ipv6EgressIPServedPodsAS)
+		qosV6Rule.ExternalIDs = getEgressIPQoSRuleDbIDs(IPFamilyValueV6).GetExternalIDs()
+		ops, err = libovsdbops.CreateOrUpdateQoSesOps(nbClient, ops, &qosV6Rule)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot create v6 QoS rule ops for egressIP feature on controller %s, %v", controllerName, err)
+		}
+		qoses = append(qoses, &qosV6Rule)
+	}
+	return qoses, ops, nil
+}
+
+func (oc *DefaultNetworkController) ensureDefaultNoRerouteQoSRules(node *v1.Node) error {
+	oc.eIPC.nodeUpdateMutex.Lock()
+	defer oc.eIPC.nodeUpdateMutex.Unlock()
+	var ops []ovsdb.Operation
+	// since this function is called from node update event, let us check
+	// libovsdb cache before trying to create insert/update ops so that it
+	// doesn't cause no-op construction spams at scale (kubelet sends node
+	// update events every 10seconds so we don't want to cause unnecessary
+	// no-op transacts that often and lookup is less expensive)
+	predicateIDs := libovsdbops.NewDbObjectIDs(libovsdbops.LogicalRouterPolicyEgressIP, DefaultNetworkControllerName,
+		map[libovsdbops.ExternalIDKey]string{
+			libovsdbops.ObjectNameKey: string(ReplyTrafficMark),
+		})
+	qosPredicate := libovsdbops.GetPredicate[*nbdb.QoS](predicateIDs, nil)
+	existingQoSes, err := libovsdbops.FindQoSesWithPredicate(oc.nbClient, qosPredicate)
+	if err != nil {
+		return err
+	}
+	qosExists := false
+	if config.IPv4Mode && config.IPv6Mode && len(existingQoSes) == 2 {
+		// no need to create QoS Rule ops; already exists; dualstack
+		qosExists = true
+	}
+	if len(existingQoSes) == 1 && ((config.IPv4Mode && !config.IPv6Mode) || (config.IPv6Mode && !config.IPv4Mode)) {
+		// no need to create QoS Rule ops; already exists; single stack
+		qosExists = true
+	}
+	if !qosExists {
+		existingQoSes, ops, err = createDefaultReRouteQoSRuleOps(oc.nbClient, oc.addressSetFactory, oc.controllerName)
```

Comment:
why do we create the ops if the node might not be local zone? could we skip calling this entire function if the node is remote?

Reply:
oh good point, my check seems too late in cycle I should probably push this to when we get the event in default network controller

Reply:
done.